### PR TITLE
Rate limit when heap usage exceeds limit [DPP-997]

### PR DESCRIPTION
**Pull request by simonmaxen-da**

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 1]
```diff
@@ -75,6 +93,35 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
+    }
+  }
+
+  /** When the collected tenured memory pool usage exceeds the threshold this state will continue even if memory
+    * has been freed up if not garbage collection takes place.  For this reason when we are over limit we also
+    * run garbage collection on every request to ensure the collection usage stats are as up to date as possible
+    * to thus stop rate limiting as soon as possible.
+    */
+  private def gc(): Unit = {
+    memoryMxBean.gc()
```
Comment (skisel-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:122]: As it basically delegates to `System.gc()` as according to `MemoryMXBean.java` I would use directly here `System.gc()` as then it clear what is done.

Checking the necessity of calling explicitly a garbage collector deserves a deep analysis on my side. There are usually good reasons exist not to ever call it explicitly :)

**Comment by simonmaxen-da** [comment]
LR passes: https://lr.da-ext.net/grafana/d/sandbox-on-x/sandbox-on-x?orgId=1&var-benchmark_id=8356&from=1654619140000&to=1654620382000

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala [hunk 2]
```diff
@@ -6,11 +6,13 @@ package com.daml.platform.apiserver.configuration
...
     maxApiServicesQueueSize: Int,
     maxApiServicesIndexDbQueueSize: Int,
+    maxHeapSpacePercentage: Int,
 )
 
 case object RateLimitingConfig {
   val default: RateLimitingConfig = RateLimitingConfig(
     maxApiServicesQueueSize = 10000,
     maxApiServicesIndexDbQueueSize = 1000,
+    maxHeapSpacePercentage = 85,
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala:16]: As discussed with @mziolekda there should also be a parameter that maxThreasholdSize.  The actual threshold used would be max(maxHeapSpacePercentage * maxHeapSize, maxHeapSize - maxThresholdSize)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 3]
```diff
@@ -75,6 +91,34 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
     }
   }
 
+  private def memoryOverloaded(fullMethodName: String): Either[String, Unit] = {
+    tenuredMemoryPool.fold[Either[String, Unit]](Right(())) { p =>
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:95]: To increase robustness of the solution I would suggest dealing explicitly with 2 possible future error cases:
* we encounter a GC that doesn't have any tenured pool
* we encounter a GC with more than one tenured pool
In both cases the mechanism should be disabled and an error message should be printed into the log

In the successful case, I would print a message saying what kind of memory pool we are hooking to and what kind of thresholds we are observing.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 4]
```diff
@@ -13,11 +13,26 @@ import io.grpc._
...
+) extends ServerInterceptor {
 
   private val logger = LoggerFactory.getLogger(getClass)
 
+  private val tenuredMemoryPool: Option[MemoryPoolMXBean] =
+    memoryPoolMxBeans.find(p =>
+      p.getType == MemoryType.HEAP && p.isCollectionUsageThresholdSupported
+    )
+
+  tenuredMemoryPool.foreach { p =>
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:32]: To increase robustness of the solution I would suggest dealing explicitly with 2 possible future error cases:
* we encounter a GC that doesn't have any tenured pool
* we encounter a GC with more than one tenured pool

In both cases the mechanism should be disabled and an error message should be printed into the log
In the successful case, I would print a message saying what kind of memory pool we are hooking to and what kind of thresholds we are observing.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala [hunk 5]
```diff
@@ -6,11 +6,24 @@ package com.daml.platform.apiserver.configuration
...
 case object RateLimitingConfig {
-  val default: RateLimitingConfig = RateLimitingConfig(
+
+  val Megabyte: Long = 1024L * 1024L
+
+  val Default: RateLimitingConfig = RateLimitingConfig(
     maxApiServicesQueueSize = 10000,
     maxApiServicesIndexDbQueueSize = 1000,
+    maxHeapSpacePercentage = 85,
+    maxOverThresholdZoneSize = 100 * Megabyte,
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala:27]: In general, I like the fact that this safety buffer is on the small-ish side of the scale. It means that for most practical installations the absolute size will be dominating over the relative one. 
Now I am starting to think that it is maybe too aggressive. We will basically leave 100MB safety buffer whenever the pool is >=667MB.

How about making it 300MB. It would mean that it will start dominating for pools >=2GB

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/GrpcServer.scala [hunk 6]
```diff
@@ -53,7 +55,16 @@ private[apiserver] object GrpcServer {
     builder.intercept(new MetricsInterceptor(metrics))
     builder.intercept(new TruncatedStatusInterceptor(MaximumStatusDescriptionLength))
     builder.intercept(new ErrorInterceptor)
-    rateLimitingConfig.foreach(c => builder.intercept(new RateLimitingInterceptor(metrics, c)))
+    rateLimitingConfig.foreach(c =>
+      builder.intercept(
+        new RateLimitingInterceptor(
+          metrics = metrics,
+          config = c,
+          memoryPoolMxBeans = ManagementFactory.getMemoryPoolMXBeans.asScala.toList,
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/GrpcServer.scala:63]: Shouldn't that be an implementation detail invisible to the users of the rate limiting interceptor? We don't need to know how the sausage is made here.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 7]
```diff
@@ -13,11 +13,26 @@ import io.grpc._
...
+private[apiserver] final class RateLimitingInterceptor(
+    metrics: Metrics,
+    config: RateLimitingConfig,
+    memoryPoolMxBeans: List[MemoryPoolMXBean],
+    memoryMxBean: MemoryMXBean,
+) extends ServerInterceptor {
 
   private val logger = LoggerFactory.getLogger(getClass)
 
+  private val tenuredMemoryPool: Option[MemoryPoolMXBean] =
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:27]: Added TenuredMemoryPool

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 8]
```diff
@@ -75,6 +83,34 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
+      if (p.isCollectionUsageThresholdExceeded) {
+        // Based on a combination of JvmMetricSet and MemoryUsageGaugeSet
+        val poolBeanMetricPrefix = s"jvm_memory_usage_pools_${p.getName}"
+        val rpcStatus =
+          s"""
+             | The ${p.getName} collection  has exceeded the maximum (${p.getCollectionUsageThreshold}).
+             | The rejected call was $fullMethodName.
+             | Jvm memory metrics are available at $poolBeanMetricPrefix
+          """.stripMargin
+        gc()
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:97]: Could you add here a comment explaining why it is safe to call gc (likely repeatedly) here?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 9]
```diff
@@ -75,6 +83,34 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
+      } else {
+        Right(())
+      }
+    }
+  }
+
+  /** When the collected tenured memory pool usage exceeds the threshold this state will continue even if memory
+    * has been freed up if not garbage collection takes place.  For this reason when we are over limit we also
+    * run garbage collection on every request to ensure the collection usage stats are as up to date as possible
+    * to thus stop rate limiting as soon as possible.
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:108]: Could you add here a comment explaining why it is safe to call gc (likely repeatedly) here?

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala [hunk 10]
```diff
@@ -164,22 +168,86 @@ final class RateLimitingInterceptorSpec
...
+        }
+    }
+  }
+
+  it should "calculate the collection threshold zone size" in {
+    // The actual threshold used would be max(maxHeapSpacePercentage * maxHeapSize / 100, maxHeapSize - maxOverThresholdZoneSize)
+    val underTest =
+      RateLimitingConfig.Default.copy(maxHeapSpacePercentage = 90, maxOverThresholdZoneSize = 1000)
+    underTest.collectionUsageThreshold(1000) shouldBe 900 // 90%
+    underTest.collectionUsageThreshold(101000) shouldBe 100000 // 10100 - 1000
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:220]:
```suggestion
    underTest.collectionUsageThreshold(101000) shouldBe 100000 // 101000 - 1000
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 11]
```diff
@@ -75,6 +83,34 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
   }
 
+  private def memoryOverloaded(fullMethodName: String): Either[String, Unit] = {
+    tenuredMemoryPool.fold[Either[String, Unit]](Right(())) { p =>
+      if (p.isCollectionUsageThresholdExceeded) {
+        // Based on a combination of JvmMetricSet and MemoryUsageGaugeSet
+        val poolBeanMetricPrefix = s"jvm_memory_usage_pools_${p.getName}"
+        val rpcStatus =
+          s"""
+             | The ${p.getName} collection  has exceeded the maximum (${p.getCollectionUsageThreshold}).
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:101]:
```suggestion
             | The ${p.getName} collection has exceeded the maximum collection usage threshold (${p.getCollectionUsageThreshold}).
```
?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 9 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:116]: Made use of a GC throttled memory bean mandatory.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 12]
```diff
@@ -75,6 +83,34 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
     }
   }
 
+  private def memoryOverloaded(fullMethodName: String): Either[String, Unit] = {
+    tenuredMemoryPool.fold[Either[String, Unit]](Right(())) { p =>
+      if (p.isCollectionUsageThresholdExceeded) {
+        // Based on a combination of JvmMetricSet and MemoryUsageGaugeSet
+        val poolBeanMetricPrefix = s"jvm_memory_usage_pools_${p.getName}"
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:98]: Is this a name of a JVM's mx bean? Or a metric name that we are setting up?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 13]
```diff
@@ -98,9 +134,65 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
+  }
+
+  val doNonLimit: Set[String] = Set(
     "grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo",
     "grpc.health.v1.Health/Check",
     "grpc.health.v1.Health/Watch",
   )
 }
+
+object TenuredMemoryPool {
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:202]: I don't think It is not general purpose enough to warrant that.

**Review by pbatko-da — COMMENTED** [review]
Have we document which GC algorithms strategies undertaken in this PR apply?
I think it'll be useful to have the supported algorithms listed explicitly somewhere in the code

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:202]: I have added some addition testing around the purpose of this class though.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 11 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:101]: Changed.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 14]
```diff
@@ -98,9 +134,65 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
+
+  private val logger = LoggerFactory.getLogger(getClass)
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:204]: That call will be done once, at startup, where there isn't really any (message) context.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 15]
```diff
@@ -98,9 +134,65 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
+
+  def apply(
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:206]: Done.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala [hunk 16]
```diff
@@ -6,11 +6,24 @@ package com.daml.platform.apiserver.configuration
 final case class RateLimitingConfig(
     maxApiServicesQueueSize: Int,
     maxApiServicesIndexDbQueueSize: Int,
-)
+    maxHeapSpacePercentage: Int,
+    maxOverThresholdZoneSize: Long,
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala:10]: Done.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 17]
```diff
@@ -103,10 +111,14 @@ private[apiserver] final class RateLimitingInterceptor(
   }
 
   /** When the collected tenured memory pool usage exceeds the threshold this state will continue even if memory
-    * has been freed up if not garbage collection takes place.  For this reason when we are over limit we also
+    * has been freed up if no garbage collection takes place.  For this reason when we are over limit we also
     * run garbage collection on every request to ensure the collection usage stats are as up to date as possible
     * to thus stop rate limiting as soon as possible.
+    *
+    * We use a throttled memory bean to ensure that even if the server is under heavy rate limited load calls
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:134]: Consider rephrasing, it is hard to follow.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:218]: I think moving it to a separate file is warranted though. We have two classes/objects with well defined fringe purpose and clean interfaces. They deserve to live on their own

**Comment by simonmaxen-da** [comment]
Added a Lab Log at:
https://digitalasset.atlassian.net/wiki/spaces/DAMLX/pages/edit-v2/2727805137

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 18]
```diff
@@ -75,6 +93,52 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
     }
   }
 
+  private def memoryOverloaded(fullMethodName: String): Either[String, Unit] = {
+    tenuredMemoryPool.fold[Either[String, Unit]](Right(())) { p =>
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:97]: How about abstracting this method and moving it to `TenuredMemoryPool`? It could just return simplified string to be further augmented with the details of `fullMethodName` outside

File: ledger/participant-integration-api/src/test/resources/logback-test.xml [hunk 19]
```diff
@@ -56,7 +56,7 @@
         <appender-ref ref="IndexerStabilitySpecAppender"/>
     </logger>
 
-    <root level="DEBUG">
+    <root level="INFO">
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/resources/logback-test.xml:59]: Is this intentional or accidental?
